Builder::SetFrameLocked is not implemented

[`Builder::SetFrameLocked`](https://github.com/jstnhuang/robot_markers/blob/e412e6b33c608ae36afccbbca15e7bda97c5be64/include/robot_markers/builder.h#L151) is defined but not implemented. It therefore produces linking errors: ``undefined reference to `robot_markers::Builder::SetFrameLocked(bool)'``.

--- a/include/robot_markers/builder.h
+++ b/include/robot_markers/builder.h
@@ -4,6 +4,7 @@
 #include <map>
 #include <set>
 #include <string>
+#include <unordered_map>
 
 #include "geometry_msgs/Pose.h"
 #include "ros/duration.h"
@@ -137,6 +138,10 @@ class Builder {
   /// \param[in] a The alpha value, between 0 and 1.
   void SetColor(float r, float g, float b, float a);
 
+  /// \brief Set link-specific color
+  /// \param[in] colors color tuple (r,g,b,a) per link
+  void SetLinkColors(std::unordered_map<std::string, std::array<float, 4> > colors);
+
   /// \brief Sets the lifetime of the robot markers.
   ///
   /// \param[in] lifetime The lifetime for the robot markers.
@@ -187,6 +192,7 @@ class Builder {
   std::string ns_;
   geometry_msgs::Pose pose_;
   std_msgs::ColorRGBA color_;
+  std::unordered_map<std::string, std_msgs::ColorRGBA> colors_;
   ros::Duration lifetime_;
   bool frame_locked_;
   bool mesh_materials_;

--- a/include/robot_markers/builder.h
+++ b/include/robot_markers/builder.h
@@ -148,7 +148,7 @@ class Builder {
   ///
   /// \param[in] frame_locked True to lock the markers to the robot's frame ID,
   ///   false otherwise.
-  void SetFrameLocked(bool frame_locked);
+  void SetFrameLocked(const bool frame_locked);
 
   /// \brief Builds a visualization of the robot model as a marker array.
   ///

--- a/include/robot_markers/builder.h
+++ b/include/robot_markers/builder.h
@@ -189,6 +189,7 @@ class Builder {
   std_msgs::ColorRGBA color_;
   ros::Duration lifetime_;
   bool frame_locked_;
+  bool mesh_materials_;
 
   bool has_initialized_;
 };

--- a/include/robot_markers/builder.h
+++ b/include/robot_markers/builder.h
@@ -171,6 +171,7 @@ class Builder {
              visualization_msgs::MarkerArray* marker_array);
 
  private:
+  void Check();
   // Sets everything in the marker except the pose.
   void BuildMarker(const urdf::Link& link, int id,
                    visualization_msgs::Marker* output);

--- a/src/builder.cpp
+++ b/src/builder.cpp
@@ -94,12 +94,7 @@ void Builder::Init() {
 
 void Builder::SetJointPositions(
     const std::map<std::string, double> joint_positions) {
-  if (!has_initialized_) {
-    ROS_ERROR(
-        "You must call Init() before calling any other methods of "
-        "robot_markers::Builder.");
-    return;
-  }
+  Check();
 
   std::vector<geometry_msgs::TransformStamped> transforms;
   fk_.GetTransforms(joint_positions, &transforms);
@@ -113,50 +108,23 @@ void Builder::SetJointPositions(
 }
 
 void Builder::SetFrameId(const std::string& frame_id) {
-  if (!has_initialized_) {
-    ROS_ERROR(
-        "You must call Init() before calling any other methods of "
-        "robot_markers::Builder.");
-    return;
-  }
-
+  Check();
   frame_id_ = frame_id;
 }
 void Builder::SetTime(const ros::Time& stamp) {
-  if (!has_initialized_) {
-    ROS_ERROR(
-        "You must call Init() before calling any other methods of "
-        "robot_markers::Builder.");
-    return;
-  }
-
+  Check();
   stamp_ = stamp;
 }
 void Builder::SetNamespace(const std::string& ns) {
-  if (!has_initialized_) {
-    ROS_ERROR(
-        "You must call Init() before calling any other methods of "
-        "robot_markers::Builder.");
-    return;
-  }
+  Check();
   ns_ = ns;
 }
 void Builder::SetPose(const geometry_msgs::Pose& pose) {
-  if (!has_initialized_) {
-    ROS_ERROR(
-        "You must call Init() before calling any other methods of "
-        "robot_markers::Builder.");
-    return;
-  }
+  Check();
   pose_ = pose;
 }
 void Builder::SetColor(float r, float g, float b, float a) {
-  if (!has_initialized_) {
-    ROS_ERROR(
-        "You must call Init() before calling any other methods of "
-        "robot_markers::Builder.");
-    return;
-  }
+  Check();
 
   color_.r = r;
   color_.g = g;
@@ -171,12 +139,7 @@ void Builder::Build(MarkerArray* marker_array) {
 
 void Builder::Build(const std::set<std::string>& link_names,
                     MarkerArray* marker_array) {
-  if (!has_initialized_) {
-    ROS_ERROR(
-        "You must call Init() before calling any other methods of "
-        "robot_markers::Builder.");
-    return;
-  }
+  Check();
 
   const std::string& root_name = model_.getRoot()->name;
 
@@ -209,6 +172,12 @@ void Builder::Build(const std::set<std::string>& link_names,
     }
     marker.pose = ToGeometryPose(transform_out);
     marker_array->markers.push_back(marker);
+  }
+}
+
+void Builder::Check() {
+  if (!has_initialized_) {
+    throw std::runtime_error("You must call Init() before calling any other methods of robot_markers::Builder.");
   }
 }
 

--- a/src/builder.cpp
+++ b/src/builder.cpp
@@ -135,6 +135,24 @@ void Builder::SetColor(float r, float g, float b, float a) {
   mesh_materials_ = false;
 }
 
+void Builder::SetLinkColors(std::unordered_map<std::string, std::array<float,4>> colors) {
+  Check();
+
+  for(const auto & [link, color] : colors) {
+      std_msgs::ColorRGBA c;
+      c.r = color[0];
+      c.g = color[1];
+      c.b = color[2];
+      c.a = color[3];
+      colors_[link] = c;
+  }
+
+  // set the default color to white
+  SetColor(1,1,1,1);
+
+  mesh_materials_ = false;
+}
+
 void Builder::SetLifetime(const ros::Duration& lifetime) {
   Check();
   lifetime_ = lifetime;
@@ -199,7 +217,7 @@ void Builder::BuildMarker(const urdf::Link& link, int id, Marker* output) {
   output->header.stamp = stamp_;
   output->ns = ns_;
   output->id = id;
-  output->color = color_;
+  output->color = colors_.count(link.name) ? colors_.at(link.name) : color_;
   output->lifetime = lifetime_;
   output->frame_locked = frame_locked_;
 

--- a/src/builder.cpp
+++ b/src/builder.cpp
@@ -132,6 +132,16 @@ void Builder::SetColor(float r, float g, float b, float a) {
   color_.a = a;
 }
 
+void Builder::SetLifetime(const ros::Duration& lifetime) {
+  Check();
+  lifetime_ = lifetime;
+}
+
+void Builder::SetFrameLocked(const bool frame_locked) {
+  Check();
+  frame_locked_ = frame_locked;
+}
+
 void Builder::Build(MarkerArray* marker_array) {
   std::set<std::string> link_names;
   Build(link_names, marker_array);

--- a/src/builder.cpp
+++ b/src/builder.cpp
@@ -66,6 +66,7 @@ Builder::Builder(const urdf::Model& model)
       color_(),
       lifetime_(0),
       frame_locked_(false),
+      mesh_materials_(true),
       has_initialized_(false) {
   pose_.orientation.w = 1;
 }
@@ -130,6 +131,8 @@ void Builder::SetColor(float r, float g, float b, float a) {
   color_.g = g;
   color_.b = b;
   color_.a = a;
+
+  mesh_materials_ = false;
 }
 
 void Builder::SetLifetime(const ros::Duration& lifetime) {
@@ -216,7 +219,7 @@ void Builder::BuildMarker(const urdf::Link& link, int id, Marker* output) {
   } else if (link.visual->geometry->type == urdf::Geometry::MESH) {
     output->type = Marker::MESH_RESOURCE;
     const urdf::Mesh& mesh = dynamic_cast<urdf::Mesh&>(*link.visual->geometry);
-    output->mesh_use_embedded_materials = true;
+    output->mesh_use_embedded_materials = mesh_materials_;
     if (mesh.scale.x == 0 && mesh.scale.y == 0 && mesh.scale.z == 0) {
       output->scale.x = 1;
       output->scale.y = 1;

